### ISSUE TITLE
Rule Change: no-useless-assignment will respect value being used as JSX Element

### ISSUE THREAD

**Issue by nikitarevenco** [eslint/eslint#18984]
### What rule do you want to change?

no-useless-assignment

### What change do you want to make?

Generate fewer warnings

### How do you think the change should be implemented?

A new default behavior

### Example code
```jsx
export function EslintDemo() {
  const Comp = "div";

  return <Comp />;
}
```

### What does the rule currently do for this code?

Generates an error on line 2

```
Diagnostics:
1. This assigned value is not used in subsequent statements. [no-useless-assignment]
```

### What will the rule do after it's changed?

It will not generate this error, because we are using the `Comp` as a JSX element

### Participation

- [ ] I am willing to submit a pull request to implement this change.

**Comment by mdjermanovic** [comment]
Hi @nikitarevenco, thanks for the issue!

Is this React, and do you have the  [`react/jsx-uses-vars`](https://github.com/jsx-eslint/eslint-plugin-react/blob/master/docs/rules/jsx-uses-vars.md) rule enabled?

As [noted in the README](https://github.com/eslint/eslint?tab=readme-ov-file#does-eslint-support-jsx), ESLint core (including the core rules) supports JSX syntax, but does not assume any semantics such as that `<Comp />` is a reference to variable `Comp`, because it's framework-specific. However, with `react/jsx-uses-vars` enabled, core rules such as `no-useless-assignment` and `no-unused-vars` should ignore variables referenced in JSX. If that's not the case, this might be a bug in the rule.

**Comment by nikitarevenco** [comment]
> Hi @nikitarevenco, thanks for the issue!
> 
> Is this React, and do you have the  [`react/jsx-uses-vars`](https://github.com/jsx-eslint/eslint-plugin-react/blob/master/docs/rules/jsx-uses-vars.md) rule enabled?
> 
> As [noted in the README](https://github.com/eslint/eslint?tab=readme-ov-file#does-eslint-support-jsx), ESLint core (including the core rules) supports JSX syntax, but does not assume any semantics such as that `<Comp />` is a reference to variable `Comp`, because it's framework-specific. However, with `react/jsx-uses-vars` enabled, core rules such as `no-useless-assignment` and `no-unused-vars` should ignore variables referenced in JSX. If that's not the case, this might be a bug in the rule.
> 

Yes this was on my end, I wasnt aware that rule needs to also be enabled. Thank you for clarifying!
